Prefer python3 over python2

pmzqla opened the pull request.

pmzqla commented:
It should be the same on OSX according to the PEP, but I don't know much about it.

What about suggesting python3 instead of python2 on Windows?

ngosang commented:
> What about suggesting python3 instead of python2 on Windows?

I can tell you from my experience that it is possible to have both versions installed in different directories but the executable is called "python" in both cases. When you run "python" in the console the executed version is the first in the PATH. I usually use python 3 and what I did was rename the executable python 2 to "python2" the thus no ambiguity. If you want to check which version is installed you have to read the Windows registry and there you can see the installation directories. Then you can call the executable with the full path "C:\\Python34\python.exe".

pmzqla commented:
I was referring to <a href="https://github.com/qbittorrent/qBittorrent/blob/master/src/gui/mainwindow.cpp#L1627">this</a> line in particular. If you have python2 installed, nothing would change, but new users could install python3.

ngosang commented:
Yes go ahead with python 3.

sledgehammer999 commented:
I don't remember how python detection happens on non-Windows systems, but on Windows we query the registry, the PATH and some default install locations under C:.
@pmzqla If you want, can you prefer python3 for all OS? (and of course change the url to the python installer too)

pmzqla commented:
@sledgehammer999 
OK, I've done some changes. I think that the `pythonSearchReg()` already returns the most recent version. I've modified the fallback detection and made it generic and updated the link to the installer.

I've no idea how things work on OSX.

Note: on Linux if both python2 or python3, but python is available, qBittorrent will claim that python is not installed. I don't think is a likely scenario, but it could happen.

ngosang commented:
> Note: on Linux if both python2 or python3, but python is available, qBittorrent will claim that python is not installed. I don't think is a likely scenario, but it could happen.

I think it's fine.

ngosang commented:
https://github.com/qbittorrent/qBittorrent/issues/3012#issuecomment-102676122
> Upgrading from Python 2.7.8 to 2.7.9 solved the problem for me. I request the sysadmins to incorporate this version check in future versions of QBittorrent.

If some search engines aren't working with Python 2.7.8 we should recommend install a new version.

sledgehammer999 commented:
Can you rebase?

pmzqla commented:
Done, but I haven't been following things that much and rebased it on master. Should I do it for v3_2_x?

sledgehammer999 commented:
Only if you want. But separate PR.

ngosang commented:
:+1: Works for me in Linux. What about show a message in the log window with the detected version?

ngosang commented:
> What about show a message in the log window with the detected version?

I mean when qBittorrents stats and detect the Python version print a message in the log. This can be useful to debug problems with search plugins.

One thought. I think most people are using Python2 at least in Windows. The problem is most unofficial plugins only works with Python2 so we can expect complains after merge this.
https://github.com/qbittorrent/qBittorrent/wiki/Unofficial-search-plugins

pmzqla commented:
>I mean when qBittorrents stats and detect the Python version print a message in the log.

I thought I had replied to this... I agree, this would be useful.

>One thought. I think most people are using Python2 at least in Windows.

I hadn't thought about unofficial plugins...

ngosang commented:
> I hadn't thought about unofficial plugins...

There are many outdated and I think they are not widely used so it's not a big problem.

sledgehammer999 commented:
Thank you. I am pushing some fixes on this too.

sledgehammer999 commented:
See https://github.com/qbittorrent/qBittorrent/commit/8b99e29dc0665cb44261d2e194d366326939afc5
